egui/src/widgets/text_edit/filter.rs

```diff
@@ -0,0 +1,178 @@
+use super::{Action, InputData};
+
+use std::{borrow::Cow, cmp};
+
+/// A type that filters user text input.
+pub trait FilterInput {
+    /// Determine how to treat the user `input` based on the current
+    /// value of `InputData`.
+    fn filter_input(data: InputData<'_>) -> Action;
+}
+
+impl FilterInput for String {
+    fn filter_input(data: InputData<'_>) -> Action {
+        Action::Insert(data.input)
+    }
+}
+
+impl FilterInput for str {
+    fn filter_input(data: InputData<'_>) -> Action {
+        Action::Insert(data.input)
+    }
+}
+
+macro_rules! impl_filter_input_unsigned {
+    ($type: ty, $max_str: expr) => {
+        impl FilterInput for $type {
+            fn filter_input(data: InputData<'_>) -> Action {
+                let InputData { buffer, input, .. } = data;
+                let max_str = $max_str;
+                let max_str_len = max_str.len();
+
+                // TODO should we assume that the user did not add data
+                // to the buffer from outside the UI?
+
+                // Filter out all the non-numeric characters.
+                let filter = input.as_str().replace(|c: char| !c.is_numeric(), "");
+
+                let new_size = buffer.len() + filter.len();
+
+                match new_size.cmp(&max_str_len) {
+                    // Guaranteed to be below the limit so we insert.
+                    cmp::Ordering::Less => Action::Insert(filter),
+                    // Guaranteed to exceed the limit so we overwrite.
+                    cmp::Ordering::Greater => Action::Overwrite(Cow::Borrowed(max_str)),
+                    // Slow case where we need to actually compare with the limit.
+                    cmp::Ordering::Equal => {
+                        // Note that we don't parse the strings because its too slow.
+                        let mut iter = buffer.chars();
+                        let front = max_str[..buffer.len()].chars();
+                        for (current, max) in (&mut iter).zip(front) {
+                            if current > max {
+                                return Action::Overwrite(Cow::Borrowed(max_str));
+                            }
+                        }
+
+                        let back = max_str[buffer.len()..].chars();
+                        for (current, max) in iter.zip(back) {
+                            if current > max {
+                                return Action::Overwrite(Cow::Borrowed(max_str));
+                            }
+                        }
+
+                        Action::Insert(filter)
```

Comment:
afaict this compares the existing buffer, but does not take into account the new input, nor where it is to be inserted.

I think we should consider an API where the user gets handed the suggested `new_buffer`, i.e. what would be produced without any filtering. The user could then very easily check that with a e.g. a regex, or lexicographically against a maximum value. In such an API we could get away with a single input (the suggested new `String`) and an `enum Action { Revert, Replace(String) }`, where `Revert` would revert to the previous value, forbidding the new input (e.g. because it doesn't fit the regex).

Presumably filters will be used for short `TextEdit`, and since the code is only run on new input we don't need to worry much about performance and should focus on usability.

Reply:
> afaict this compares the existing buffer, but does not take into account the new input, nor where it is to be inserted.

This basically always assumes the worst case scenario that the user has single cursor (and thus nothing selected). So this won't work properly when strings are selected indeed.

> I think we should consider an API where the user gets handed the suggested new_buffer, i.e. what would be produced without any filtering.

Yeah that sounds like a more user friendly API.